packages/fuels-programs/src/call_utils.rs

```diff
@@ -334,24 +357,51 @@ pub(crate) fn get_single_call_instructions(
     if let Some(inner_type_byte_size) = output_param_type.heap_inner_element_size() {
-        instructions.extend([
-            // The RET register contains the pointer address of the `CALL` return (a stack
-            // address).
-            // The RETL register contains the length of the `CALL` return (=24 because the
-            // Vec/Bytes/String struct takes 3 WORDs).
-            // We don't actually need it unless the Vec/Bytes/String struct encoding changes in the
-            // compiler.
-            // Load the word located at the address contained in RET, it's a word that
-            // translates to a heap address. 0x15 is a free register.
-            op::lw(0x15, RegId::RET, 0),
-            // We know a Vec/Bytes/String struct has its third WORD contain the length of the
-            // underlying vector, so use a 2 offset to store the length in 0x16, which is a free
-            // register.
-            op::lw(0x16, RegId::RET, 2),
-            // The in-memory size of the type is (in-memory size of the inner type) * length
-            op::muli(0x16, 0x16, inner_type_byte_size as u16),
-            op::retd(0x15, 0x16),
-        ]);
+        if let ParamType::Enum { variants, .. } = output_param_type {
+            // This offset is here because if the variant type is an enum containing a heap type,
+            // the 3 words containing (ptr, cap, len) are at the end of the data of the enum, which
+            // depends on the encoding width. If the variant type does not contain a heap type, then
+            // the receipt generated will not be read anyway, so we can generate a receipt with
+            // empty return data. 3 is not a "magic" number: we are looking left of the 3 words
+            // (ptr, cap, len), that are placed at the end of the data, on the right.
+            let offset = (output_param_type.compute_encoding_width() - 3) as u16;
+            let selected_discriminant = variants.heap_type_variant_discriminant().unwrap() as u16;
+            instructions.extend([
+                // All the registers 0x15-0x18 are free
+                // The RET register contains the pointer address of the `CALL` return (a stack
+                // address).
+                // The RETL register contains the length of the `CALL` return (=24 in the case of
+                // Vec/Bytes/String because their struct takes 3 WORDs). We don't actually need it
+                // unless the Vec/Bytes/String struct encoding changes in the compiler.
+                // Load the word located at the address contained in RET, it's a word that translates to
+                // a heap address. We use the offset because in the case of an enum containing a heap
+                // type, we need to get the pointer that is located at the heap address pointed to by
+                // the RET register.
+                op::lw(0x15, RegId::RET, offset),
```

Comment:
This load must be hidden behind the `jnef` check below.

Consider the following sway fn:
```sway
    fn a_problem() -> Result<Bytes, b256> {
        Result::Err(0x1111111111111111111111111111111111111111111111111111111111111111)
    }
```

This will cause a `MemoryOverflow` panic when executed. 

This is because we read the memory here without checking whether the discriminant is pointed towards the heap type. If it is not then we're reading whatever *is* there as a ptr.

This wasn't caught by tests because the error types used, when interpreted as a ptr, pointed to allocated memory. In order to showcase the bug I made the error a `b256` with a high value, so that we will index out-of-bounds memory.